update from gym to gymnasium

Alex2782 opened the pull request:
I haven't tested everything yet: 'stocks-v0', render_mode="human" works already.

There are still 2 warnings:
```
/opt/homebrew/lib/python3.11/site-packages/gymnasium/utils/passive_env_checker.py:35: 
UserWarning: WARN: A Box observation space has an unconventional shape (neither an image, nor a 1D vector). 
We recommend flattening the observation to have only a 1D vector 
or use a custom policy to properly process the data. Actual observation shape: (10, 2)

/opt/homebrew/lib/python3.11/site-packages/gymnasium/utils/passive_env_checker.py:203: 
UserWarning: WARN: Future gymnasium versions will require that `Env.reset` can be passed `options` 
to allow the environment initialisation to be passed additional information.
```

Alex2782 committed:
update from gym to gymnasium

elliottower commented:
The warning for flattened observations can be ignored, more of a best practices for RL algorithms thing rather than an actual requirement. 

Do you know what options are passed to reset()? If possible it’s best to change that for future compatibility, as the message says. Seeding can be done via the reset(seed) argument (directly calling seed() is deprecated)

Alex2782 committed:
part 2, update from gym to gymnasium

Alex2782 commented:
In the 2nd part I tried "reinforce"-example with 'gym-anytrading' and added more changes.

https://gymnasium.farama.org/tutorials/training_agents/reinforce_invpend_gym_v26/

In the next days I try to insert more examples and check more precisely some calculations 'max_possible_profit' / 'avg_profit'. 'total_reward' is often positive but 'profit' is less than 1 (profit loss?)

Alex2782 committed:
Create SB3_quantstats.ipynb

Alex2782 commented:
I try to take 'observation' and 'rewards' functions from 'gym-mtsim', with 'SB3' the models are not trained properly.

The 'evaluate' and 'predict' functions always return the same value (all the time only 0 or 1) 
I also tried in a virtual environment, original versions **without** 'gymnasium' support, the results were identical. 

https://github.com/DLR-RM/stable-baselines3/pull/1327 (+ 'gymnasium' support)

I could no longer successfully install 'stable_baselines' (as in the example https://github.com/AminHP/gym-anytrading/blob/master/examples/a2c_quantstats.ipynb) under macOS 13 for comparison.


### 'SB3'-predict:  A2C or PPO
https://github.com/Alex2782/gym-anytrading/blob/master/examples/SB3_quantstats.ipynb

![sb3_bug](https://user-images.githubusercontent.com/41921395/227020341-5cc2ca76-5c1d-47e1-95e6-fb88307b21b9.png)

Alex2782 committed:
adjusted 'observation_space' for SB3

Alex2782 commented:
**'observation_space'** is now initiated with **'INF = 1e10',** like in the **'gym-mtsim'** project (no longer with **'np.inf'**)
[gym_anytrading/envs/trading_env.py](https://github.com/AminHP/gym-anytrading/pull/86/commits/7288a1e3f7089b477caf846ddcc80b60c3829b7c#diff-5d3f71bdaa90f138b62b611d8a6a0e90090f893152c2e313975a7bd6c43d8238R38)

'SB3'-predict: A2C looks better now
https://github.com/Alex2782/gym-anytrading/blob/master/examples/SB3_quantstats.ipynb
![Bildschirm­foto 2023-03-25 um 00 55 03](https://user-images.githubusercontent.com/41921395/227663786-004842f7-caa0-4419-b220-29809301fad9.png)

elliottower commented:
Cool to see you got it working!

Alex2782 committed:
Create train_SB3_gymnasium.py

Alex2782 commented:
new example: [train_SB3_gymnasium.py](https://github.com/AminHP/gym-anytrading/pull/86/commits/ee37553136c8d3de64e28b2e070d53ad350af9df)

Avg. Rewards, **Supported SB3 Agents** x  learning_timesteps = [5K, 10K, 25K], Comparison with "Random actions"

![anytrading_sb3_agents](https://user-images.githubusercontent.com/41921395/228015847-13c02308-94d6-4505-8349-d7a49ef8caab.png)

AminHP commented:
Hi guys. Apologize for my delayed participation in this PR. Thanks @Alex2782 for the PR, the changes are neat and concise. The additional examples are also quite useful, the project needed more examples with newer versions of stable-baselines.

Please let me know whenever you have finished. I'm looking forward to reviewing and testing it.

Alex2782 commented:
Hi @AminHP, 

3 weeks ago I could train successfully with "REINFORCE": https://github.com/Alex2782/gym-anytrading/blob/master/examples/train_REINFORCE.py 


"stable release" for SB3 with "gymnasium" support has not been released yet.
https://github.com/DLR-RM/stable-baselines3/pull/1327#issuecomment-1508626600

AminHP commented:
Hi @Alex2782 , it seems they have just released a [new version](https://github.com/DLR-RM/stable-baselines3/releases/tag/v2.0.0) of stable-baselines3 with Gymnasium support. Can we proceed based on that?

elliottower commented:
For what it’s worth we would love to have some sample code using gymnasium and SB3 for making a tutorial, if you can get it working let me know (I haven’t had time to mess with it myself)

kmafutah commented:
Had you seen this

https://github.com/AminHP/gym-anytrading/issues/87
That’s where I got sample code from, not sure is this is the same SB3 you
talking about though.

Hope that helps

Kind regards

On Sun, 25 Jun 2023 at 19:31, Elliot Tower ***@***.***> wrote:

> For what it’s worth we would love to have some sample code using gymnasium
> and SB3 for making a tutorial, if you can get it working let me know (I
> haven’t had time to mess with it myself)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AminHP/gym-anytrading/pull/86#issuecomment-1606174015>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA7JXSYECC5XR4HGEZEJZKTXNBYV3ANCNFSM6AAAAAAWALMSZA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 


Kudakwashe Mafutah
[image: https://]about.me/kmafutah
<https://about.me/kmafutah?promo=email_sig&utm_source=product&utm_medium=email_sig&utm_campaign=edit_panel&utm_content=plaintext>

elliottower commented:
Makes sense, I realized they actually updated their entire documentation
website so now everything is gymnasium, therefore we can just link to that

On Sun, Jun 25, 2023 at 5:58 PM Kudakwashe Mafutah ***@***.***>
wrote:

> Had you seen this
>
> https://github.com/AminHP/gym-anytrading/issues/87
> That’s where I got sample code from, not sure is this is the same SB3 you
> talking about though.
>
> Hope that helps
>
> Kind regards
>
> On Sun, 25 Jun 2023 at 19:31, Elliot Tower ***@***.***> wrote:
>
> > For what it’s worth we would love to have some sample code using
> gymnasium
> > and SB3 for making a tutorial, if you can get it working let me know (I
> > haven’t had time to mess with it myself)
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/AminHP/gym-anytrading/pull/86#issuecomment-1606174015>,
>
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AA7JXSYECC5XR4HGEZEJZKTXNBYV3ANCNFSM6AAAAAAWALMSZA>
>
> > .
> > You are receiving this because you are subscribed to this thread.Message
> > ID: ***@***.***>
> >
> --
>
>
> Kudakwashe Mafutah
> [image: https://]about.me/kmafutah
> <
> https://about.me/kmafutah?promo=email_sig&utm_source=product&utm_medium=email_sig&utm_campaign=edit_panel&utm_content=plaintext>
>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AminHP/gym-anytrading/pull/86#issuecomment-1606275983>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHVPVAZ52WQZWPKYVEWBWF3XNCX7BANCNFSM6AAAAAAWALMSZA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Alex2782 commented:
Hi @AminHP,

maybe next week or in 2 weeks I have more time. 
Should I better create a new PULL request with fewer examples?

Alex2782 committed:
Update train_SB3_gymnasium.py

Alex2782 commented:
Hi @AminHP 

I have tried SB3 release version: `pip install 'stable-baselines3[extra]'`
A2C and PPO work so far, after 25K learning steps they reach more rewards than "Random Actions"

**train_sb3_gymnasium.py**

![Bildschirmfoto 2023-07-04 um 22 58 39](https://github.com/AminHP/gym-anytrading/assets/41921395/1cf74a46-0250-4b8f-bc7e-27a3e72b06d9)

AminHP commented:
Hi @Alex2782 , great job! Thanks for the update. I will review it soon and we will discuss it.

AminHP commented on gym_anytrading/envs/trading_env.py:
```diff
@@ -21,25 +20,28 @@ def opposite(self):
         self.window_size = window_size
         self.prices, self.signal_features = self._process_data()
-        self.shape = (window_size, self.signal_features.shape[1])
+        self.shape = (window_size * 2, )
```
This change seems not necessary.

AminHP commented on gym_anytrading/envs/trading_env.py:
```diff
@@ -21,25 +20,28 @@ def opposite(self):
         self.window_size = window_size
         self.prices, self.signal_features = self._process_data()
-        self.shape = (window_size, self.signal_features.shape[1])
+        self.shape = (window_size * 2, )
```
This change seems not necessary. I was able to run the examples with a 2d shape.

Alex2782 replied:
yes probably not necessary if this warning or recommendation does not matter

`utils/passive_env_checker.py:42: UserWarning: WARN: A Box observation space has an unconventional shape (neither an image, nor a 1D vector). We recommend flattening the observation to have only a 1D vector or use a custom policy to properly process the data. Actual observation shape: (30, 2)`

AminHP commented on gym_anytrading/envs/trading_env.py:
```diff
@@ -49,14 +51,23 @@ def __init__(self, df, window_size):
         self.history = None
 
+        # create cache for faster training
+        self._observation_cache = []
```
Does this cache really improve the speed? It only stores multiple slices of self.signal_features. I haven't tested the speed but I don't think it is slow to read a slice of self.signal_features in each step. Also, self.signal_features is calculated once in the class constructor and somehow is already cached.

Alex2782 replied:
not necessary for 2D shape.

for 1D you have to run the function ".flatten()", this reduced the performance a bit. (maybe only 1 to 2 %)

AminHP commented:
I have tested all the examples and everything works fine. Thanks again for your contribution and effort.

About the examples, I agree with you. I also think it is better to use jupyter notebook for the examples instead of python scripts, so the users would comprehend the examples more easily. Besides, we can merge the presented codes in this PR and provide two examples: `a2c_quantstats.ipynb` (which is the updated version of my code as you already have implemented in SB3_quantstats.ipynb), and `reinforce.ipynb` (which trains and evaluates multiple RL algorithms). Note that it seems the `random_actions` example is not necessary as it is available in the README file.

What's your opinion @Alex2782 ?

Alex2782 committed:
Update trading_env.py

- 2d shape
- removed _observation_cache

Alex2782 committed:
Examples deleted

Alex2782 commented:
Hi @AminHP,

i changed to 2d shape and removed `_observation_cache`.

4 months ago I first trained the changes with `REINFORCE` example. The example did not work with 2d shape.
https://gymnasium.farama.org/tutorials/training_agents/reinforce_invpend_gym_v26/

`random_actions` example and `REINFORCE` example I have also removed.

What exactly I did to the "ipynb" files, I unfortunately don't remember, after execution the file was completely overwritten.

Alex2782 commented:
@AminHP 
the `render` function has been revised
https://gymnasium.farama.org/content/migration-guide/#environment-render

I am not sure if more changes are necessary.

In the project 'gym-anytrading' it can still be executed if 'render_mode' is configured correctly. 

In the project 'gym-mtsim' it was not compatible at all yet
https://github.com/AminHP/gym-mtsim/pull/40

AminHP commented on examples/train_SB3_gymnasium.py:
```diff
@@ -0,0 +1,211 @@
+# stable-baselines3
```
Could you create an ipynb version of this example, please?

Alex2782 replied:
ok I try to find time in the next few days

AminHP replied:
Great, thanks!

Alex2782 replied:
`train_SB3_gymnasium.ipynb`
https://github.com/AminHP/gym-anytrading/blob/3f7192b3734d399de526f2e0252adbae6ffd6e99/examples/train_SB3_gymnasium.ipynb

AminHP commented on examples/SB3_quantstats.ipynb:
```diff
@@ -0,0 +1,634 @@
+{
```
Please also add SB3_quantstats.html and remove quantstats-tearsheet.html

Alex2782 replied:
`qs.reports.html(returns, output='SB3_quantstats.html')`

but the file 'quantstats-tearsheet.html' is created again, why?
I copied your a2c example 4 months ago and adjusted only a few lines. 

I am not familiar with 'quantstats' at all. The graphics look identical. Better also remove 'SB3_quantstats.ipynb'?

AminHP replied:
I think 'SB3_quantstats.ipynb' is a good example, which was requested by one the users when I started this project, and we should keep it. About the 'quantstats-tearsheet.html' problem, I can work on it later if you haven't fixed it yet.

Alex2782 replied:
I have updated your a2c example, no longer compatible with SB3.

'quantstats' has a parameter for filename. Probably new version necessary.
https://github.com/ranaroussi/quantstats/blob/main/quantstats/reports.py#L63

Alex2782 committed:
Update a2c_quantstats

Alex2782 committed:
Create train_SB3_gymnasium.ipynb

AminHP commented:
I have tested the examples and everything works fine. Thanks @Alex2782  for patiently addressing and resolving the matters. Congratulations on your first contribution to this repo!

AminHP commented:
I will update the docs and make some minor changes later.
